File: app/src/main/java/com/example/neon/deardiary/diaryedit/DiaryEditPresenter.java
package com.example.neon.deardiary.diaryedit;

import com.example.neon.deardiary.data.Diary;
import com.example.neon.deardiary.data.source.DataSource;


class DiaryEditPresenter implements DiaryEditContract.Presenter {
    private DiaryEditContract.View mView;
    private DataSource mDataSource;
//    private static final String TAG = "DiaryEditPresenter";

    DiaryEditPresenter(DataSource dataSource, DiaryEditContract.View view) {
        mDataSource = dataSource;
        mView = view;
        mView.setPresenter(this);
    }

    @Override
    public void start() {

    }

    @Override
    public void saveDiary(Diary diary) {
        mDataSource.updateDiary(diary, new DataSource.UpdateDiaryCallback() {
            @Override
            public void onDiaryUpdated() {
                mView.onDiaryUpdated();
            }

            @Override
            public void onDiaryUpdateFailed() {
                mView.onDiaryUpdateFailed();
            }
        });

    }

    @Override
    public int getValidDiaryCount() {
        return mDataSource.getValidDairyCount();
    }

}

File: app/src/main/java/com/example/neon/deardiary/diaryedit/DiaryEditPresenter.kt
package com.example.neon.deardiary.diaryedit

import com.example.neon.deardiary.data.Diary
import com.example.neon.deardiary.data.source.DataSource


internal class DiaryEditPresenter(private val mDataSource: DataSource?, private val mView: DiaryEditContract.View) : DiaryEditContract.Presenter {

    init {
        mView.setPresenter(this)
    }

    override fun start() {

    }

    override fun validDiaryCount(): Int = mDataSource !!.validDairyCount()

    override fun saveDiary(diary: Diary) {
        mDataSource !!.updateDiary(diary, object : DataSource.UpdateDiaryCallback {
            override fun onDiaryUpdated() {
                mView.onDiaryUpdated()
            }

            override fun onDiaryUpdateFailed() {
                mView.onDiaryUpdateFailed()
            }
        })

    }
}
